_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,1468 @@
+  def sendAndReceive(uri: Uri, requestOpt: Option[Request[IO]]): IO[String]
+
+  object HttpClient:
+
+    def of(client: Client[IO]): HttpClient = new HttpClient:
+      override def sendAndReceive(uri: Uri, requestOpt: Option[Request[IO]]): IO[String] =
+        requestOpt match
+          case Some(request) => client.expect[String](request.withUri(uri))
+          case None          => client.expect[String](uri)
+```
+
+It's worthwile to mention, that:
+- `def of(client: Client[IO]): HttpClient:` - This is a factory method that creates an instance of `HttpClient`. It takes a `Client[IO]` as a parameter, which is presumably an HTTP client provided by the `http4s` library. It returns a new `HttpClient` instance.
+
+- `override def sendAndReceive(uri: Uri, requestOpt: Option[Request[IO]]): IO[String]`: This method overrides the `sendAndReceive` method declared in the `HttpClient` trait. It takes a `Uri` and an optional `Request[IO]`. Depending on whether a request is provided `(Some(request))` or not `(None)`, it uses the `client.expect[String]` method to send an HTTP request and receive a response. The response body is expected to be of type `String`, and it is returned wrapped in an `IO` monad.  If a request is provided `(Some(request))`, it sends the request with the specified uri.  If no request is provided `(None)`, it just pings the specified uri directly.
+
+Now that we have `HttpClient` in place we can move on and implement more interesting services, such as `SendAndExpect[A]` which directly uses `HttpClient`.
+
+But before we do so we need to define the data type for understanding whether the server is alive or not, this will be pretty straightforward:
+```scala
+package com.rockthejvm.loadbalancer.http
+
+enum ServerHealthStatus:
+  case Alive, Dead
+```
+
+Now, back to `SendAndExpect[A]`,
+
+This will be a special service which:
+- forwards request to the backend, receives response and returns it to the client
+- pings the backend to check its health
+
+Let's translate this to code:
+```scala
+
+import com.rockthejvm.loadbalancer.http.{HttpClient, ServerHealthStatus}
+import org.http4s.client.UnexpectedStatus
+import org.http4s.{Request, Uri}
+import cats.syntax.option.*
+import cats.effect.IO
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+import org.typelevel.log4cats.syntax.*
+import cats.syntax.applicative._
+
+import scala.concurrent.duration.DurationInt
+
+trait SendAndExpect[A]:
+  def apply(uri: Uri): IO[A]
+
+object SendAndExpect:
+
+  implicit def logger: Logger[IO] = Slf4jLogger.getLogger[IO]
+
+    def toBackend(httpClient: HttpClient, req: Request[IO]): SendAndExpect[String] =
+      new SendAndExpect[String]:
+        override def apply(uri: Uri): IO[String] =
+          info"[LOAD-BALANCER] sending request to $uri" *> httpClient
+            .sendAndReceive(uri, req.some)
+            .handleErrorWith:
+                case UnexpectedStatus(org.http4s.Status.NotFound, _, _) =>
```

Comment:
This is one of the places where I'm really not fond of the braceless Scala 3 syntax.
But at least you're consistent throughout the article, and that makes it ok.